source/Weather.cpp

```diff
@@ -70,29 +71,50 @@ const Point &Weather::Origin() const
 
 
 // Create any environmental effects and decrease the lifetime of this weather.
-void Weather::Step(vector<Visual> &visuals)
+void Weather::Step(vector<Visual> &visuals, const Point &center)
 {
 	// Environmental effects are created by choosing a random angle and distance from
 	// their origin, then creating the effect there.
 	double minRange = hazard->MinRange();
 	double maxRange = hazard->MaxRange();
+	double effectMultiplier = currentStrength;
 
-	// Estimate the number of visuals to be generated this frame.
-	// MAYBE: create only a subset of possible effects per frame.
-	int totalAmount = 0;
-	for(auto &&effect : hazard->EnvironmentalEffects())
-		totalAmount += effect.second;
-	totalAmount *= currentStrength;
-	visuals.reserve(visuals.size() + totalAmount);
-
-	for(auto &&effect : hazard->EnvironmentalEffects())
-		for(int i = 0; i < effect.second * currentStrength; ++i)
-		{
-			Point angle = Angle::Random().Unit();
-			double magnitude = (maxRange - minRange) * sqrt(Random::Real());
-			Point pos = origin + (minRange + magnitude) * angle;
-			visuals.emplace_back(*effect.first, std::move(pos), Point(), Angle::Random());
-		}
+	// If a hazard is system-wide, the max range becomes the edge of the screen,
+	// and the number of effects drawn is scaled accordingly.
+	if(hazard->SystemWide())
+	{
+		// Find the farthest possible point from the screen center and use that as
+		// our new max range. Multiply by 2 to account for the max view zoom level.
+		double newMax = 2. * Screen::Dimensions().Length();
+		// Maintain the same density of effects by dividing the new area
+		// by the old. (The pis cancel out and therefore need not be taken
+		// into account.)
+		effectMultiplier *= (newMax * newMax) / (maxRange * maxRange);
+		maxRange = newMax;
```

Comment:
`effectMultiplier` will be updated for the very first frame only. Is this intended? If we want to maintain the same density don't we have to do this every frame?

Reply:
>`effectMultiplier` will be updated for the very first frame only.

Doesn't look like it to me. Weather::Step is called for every frame that the Weather is alive. The effectMultiplier variable is created at the start of the function. Hazard is const and therefore hazard->SystemWide() never changes its truth value. Therefore, this line always gets run for every frame that a system-wide hazard is alive.

Reply:
My bad, I didn't realize that `maxRange` is a local variable.